canopen/profiles/p402.py

```diff
@@ -45,59 +201,218 @@ class Node402(RemoteNode):
-    		# corresponds with the value[1] to determine the current status
-            bitmaskvalue = statusword & value[0]
-            if bitmaskvalue == value[1]:
-                mapobject.pdo_node.node.powerstate_402._state = key
+        """
+        try:
+            logger.info('Changing Operation Mode to {0}'.format(mode))
+            state = self.state
+            result = False
+
+            if not self.is_op_mode_supported(mode):
+                raise TypeError('Operation mode not suppported by the node.')
+
+            if self.state == 'OPERATION ENABLED':
+                self.state = 'SWITCHED ON'
+                # to make sure the node does not move with a old value in another mode
+                # we clean all the target values for the modes
+                self.sdo[0x60FF].raw = 0.0  # target velocity
+                self.sdo[0x607A].raw = 0.0  # target position
+                self.sdo[0x6071].raw = 0.0  # target torque
+            # set the operation mode in an agnostic way, accessing the SDO object by ID
+            self.sdo[0x6060].raw = OperationMode.NAME2CODE[mode]
+            t = time.time() + 0.5  # timeout
+            while self.op_mode != mode:
+                if time.time() > t:
+                    raise RuntimeError('Timeout setting the new mode of operation at node {0}.'.format(self.id))
+            result = True
+        except SdoCommunicationError as e:
+            logger.warning('[SDO communication error] Cause: {0}'.format(str(e)))
+        except (RuntimeError, ValueError) as e:
+            logger.warning('{0}'.format(str(e)))
+        finally:
+            self.state = state  # set to last known state
+            logger.info('Mode of operation of the node {n} is {m}.'.format(n=self.id , m=mode))
+        return result
+
+    def is_op_mode_supported(self, mode):
+        """Function to check if the operation mode is supported by the node
+        :param int mode: Operation mode
+        :return: If the operation mode is supported
+        :rtype: bool
+        """
+        mode_support = (self.sdo[0x6502].raw & OperationMode.SUPPORTED[mode])
+        return mode_support == OperationMode.SUPPORTED[mode]
+
+    def on_TPDOs_update_callback(self, mapobject):
+        """This function receives a map object.
+        this map object is then used for changing the
+        :param mapobject: :class: `canopen.objectdictionary.Variable`
+        """
+        for obj in mapobject:
+            self.tpdo_values[obj.index] = obj.raw
+
+    @property
+    def statusword(self):
+        """Returns the last read value of the Statusword (0x6041) from the device.
+        :raise ValueError: The Object 0x6041 (Statusword) is not configured in this device.
+        """
+        try:
+            return self.tpdo_values[0x6041]
+        except KeyError:
+            raise KeyError('The object 0x6041 (Statusword) is not configured in this device.')
+
+    @statusword.setter
+    def statusword(self, value):
```

Comment:
This error is annoying but i'm not seeing a way to remove it, the setter can't have the  ``@classmethod`` decorator so... 
My intent is to throw the exception if setting a statusword. I set the statusword as a property because I think is a good design choice.

Everything works.

@christiansandberg if you want, you can cherry pick the commit for the renamed ``.msg`` to ``.pdo_parent`` or if you agree this is a good approach for the ds402 profile you can accept the PR. If anyone have suggestion (you included) or bugs to report I'm glad to solve them.

Thanks

Reply:
I think you can remove the methods that don’t do anything. Then the user should simply get an exception that the attribute does not exist or is not settable.